src/sema/diagnostics.rs

```diff
@@ -34,17 +34,23 @@ impl Diagnostics {
     }
 
     pub fn push(&mut self, diagnostic: Diagnostic) {
-        if matches!(diagnostic.level, Level::Error) {
-            self.has_error = true;
+        if !self.contents.contains(&diagnostic) {
+            if matches!(diagnostic.level, Level::Error) {
+                self.has_error = true;
+            }
+            self.contents.push(diagnostic);
         }
-        self.contents.push(diagnostic);
     }
 
     pub fn extend(&mut self, diagnostics: Vec<Diagnostic>) {
         if !self.has_error {
             self.has_error = diagnostics.iter().any(|m| m.level == Level::Error);
         }
-        self.contents.extend(diagnostics);
+        for diagnostic in diagnostics {
+            if !self.contents.contains(&diagnostic) {
```

Comment:
You are searching the entire vector for each new diagnostic!
This search should be improved to a hash table or a binary search.